libsignal-service/src/sealed_session_cipher.rs

```diff
@@ -0,0 +1,1070 @@
+            chain_key: ephemeral_derived[0..32].into(),
+            cipher_key: ephemeral_derived[32..64].into(),
+            mac_key: ephemeral_derived[64..96].into(),
+        };
+        Ok(ephemeral_keys)
+    }
+
+    fn calculate_static_keys(
+        &self,
+        public_key: &PublicKey,
+        private_key: &PrivateKey,
+        salt: &[u8],
+    ) -> Result<StaticKeys, SealedSessionError> {
+        let static_secret = public_key.calculate_agreement(private_key)?;
+        let static_derived = libsignal_protocol::create_hkdf(&self.context, 3)?
+            .derive_secrets(96, &static_secret, salt, &[])?;
+        Ok(StaticKeys {
+            cipher_key: static_derived[32..64].into(),
+            mac_key: static_derived[64..96].into(),
+        })
+    }
+
+    fn encrypt_bytes(
+        &self,
+        cipher_key: &[u8],
+        mac_key: &[u8],
+        plaintext: &[u8],
+    ) -> Result<Vec<u8>, SealedSessionError> {
+        let mut cipher = Aes256Ctr::new(cipher_key.into(), &[0u8; 16].into());
+
+        let mut ciphertext = plaintext.to_vec();
+        cipher.encrypt(&mut ciphertext);
+
+        let mut mac = Hmac::<Sha256>::new_varkey(&mac_key)
+            .map_err(|_| MacError::InvalidKeyLength)?;
+        mac.update(&ciphertext);
+        let our_mac = mac.finalize().into_bytes();
+
+        let encrypted = [ciphertext.as_slice(), &our_mac[..10]].concat();
+
+        Ok(encrypted)
+    }
+
+    fn decrypt_bytes(
+        cipher_key: &[u8],
+        mac_key: &[u8],
+        ciphertext: &[u8],
+    ) -> Result<Vec<u8>, SealedSessionError> {
+        if ciphertext.len() < 10 {
+            return Err(SealedSessionError::InvalidMacError(
+                MacError::CiphertextNotLongEnough(ciphertext.len()),
+            ));
+        }
+
+        let (ciphertext_part1, their_mac) =
+            ciphertext.split_at(ciphertext.len() - 10);
+
+        let mut verifier = Hmac::<Sha256>::new_varkey(&mac_key)
+            .map_err(|_| MacError::InvalidKeyLength)?;
+        verifier.update(&ciphertext_part1);
+        let digest = verifier.finalize().into_bytes();
+        let our_mac = &digest[..10];
+
+        if our_mac != their_mac {
+            return Err(SealedSessionError::InvalidMacError(MacError::BadMac));
+        }
+
+        let mut decrypted = ciphertext_part1.to_vec();
+        let mut cipher = Aes256Ctr::new(cipher_key.into(), &[0u8; 16].into());
+        cipher.decrypt(&mut decrypted);
+
+        Ok(decrypted)
+    }
+
+    fn get_preferred_address(
+        &self,
+        certificate: &SenderCertificate,
+    ) -> Result<Address, SealedSessionError> {
+        if let Some(ref sender_uuid) = certificate.sender_uuid {
+            let address =
+                Address::new(sender_uuid, certificate.sender_device_id as i32);
+            if self.store_context.contains_session(&address)? {
+                return Ok(address);
+            }
+        } else if let Some(ref sender_e164) = certificate.sender_e164 {
```

Comment:
`else` shouldn't go here.  The logic is the same as in the other cipher, can this be somehow refactored?

Reply:
bug fixed, thanks! I shared the implementation of `get_preferred_protocol_address` with the `SessionCipher`.